metadata-ingestion/src/datahub/ingestion/source/state/entity_removal_state.py

```diff
@@ -7,19 +7,32 @@
     StaleEntityCheckpointStateBase,
 )
 from datahub.utilities.checkpoint_state_util import CheckpointStateUtil
+from datahub.utilities.dedup_list import deduplicate_list
 from datahub.utilities.urns.urn import guess_entity_type
 
 
 class GenericCheckpointState(StaleEntityCheckpointStateBase["GenericCheckpointState"]):
     urns: List[str] = pydantic.Field(default_factory=list)
 
+    # We store a bit of extra internal-only state so that we can keep the urns list deduplicated.
+    # However, we still want `urns` to be a list so that it maintains its order.
+    # We can't used OrderedSet here because pydantic doesn't recognize it and
+    # it isn't JSON serializable.
+    _urns_set: set = pydantic.PrivateAttr(default_factory=set)
+
+    def __init__(self, **data: Any):  # type: ignore
+        super().__init__(**data)
+        self.urns = deduplicate_list(self.urns)
+        self._urns_set = set(self.urns)
+
     @classmethod
     def get_supported_types(cls) -> List[str]:
         return ["*"]
 
     def add_checkpoint_urn(self, type: str, urn: str) -> None:
-        # TODO: dedup
-        self.urns.append(urn)
+        if urn not in self._urns_set:
+            self.urns.append(urn)
+            self._urns_set.add(urn)
```

Comment:
Do we need the set if we deduplicate the list in the beginning and we only add new items to the list?

Reply:
We're not guaranteed that we only add new items to the list once - for example, I think the dbt source sometimes adds urns to the state multiple times.

Really the deduplication on initialization isn't necessary, since our state payloads already shouldn't have any duplicates. However, I still added it because it's cheap and improves code clarity.